### PR TITLE
Match `@types/airtable` types and fix type lint warnings

### LINKED ISSUE #210: Typescript return type for table.Create() does not overload correctly
Hey team,

I am updating my code to use promises rather than callbacks and have just come across this error, essentially when using the create function and passing it an array I am not getting returned a corresponding array of Records:

image

I can see in the typings this is meant to be the case:

Create method signature
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/airtable/index.d.ts#L46

Records type:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/airtable/index.d.ts#L80

For now I am just going to drop a @ts-ignore on this. But I wanted to feed this back to the team.

### LINKED ISSUE #220: Included typings in 0.10.0 are less useful than the previous DT typings
Thanks for moving the TS definitions into this library in 0.10.0!

Unfortunately, while the [DT definitions for 0.8.0](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/bacc18855fdebeec1d6edb68677100684254fa46/types/airtable/index.d.ts) expose many useful typings for various concepts in the library, such as `Base`, `Table`, `Record`, etc., the included typings in 0.10.0 only expose the top-level `Airtable` class.

Also, whereas the DT typings are generic where it makes sense (for example, `Table` and `Record` are generic on `TFields`), the new typings are significantly looser and don't provide any way to parameterize their function calls.

Both of these differences make the new typings significantly less useful to consumers. For example, it's no longer possible to type something like this without some significant workarounds, because `Airtable.Record` is no longer exposed (as a named export nor as a namespace property), and because the internal `Record` type is no longer generic:

```ts
function getFooFromRecord(record: Airtable.Record<{foo: number}>): number {
  return record.fields.foo
}

const records = await new Airtable().base('...').select().all()
const foos = records.map(getFooFromRecord)
```

Would you consider revising the published types to align with the functionality provided by the DT types? Alternatively, if this is too large a task, maybe consider removing the included types from the published package so that consumers can continue to depend on the DT typings.

### PR DESCRIPTION
**Pull request by mzgoddard**
This PR fixes a few issues related to the exported typescript types and any lint warnings.

- Fix #220 
- Fix #210 

## Improve exported Types

This PR adds exports a namespace and many types like [`@types/airtable`].

### /src/airtable.ts

The `airtable.ts` file exports the Airtable class and namespace that will be imported from this package. At the bottom of the file the Airtable namespace exports the types like those in [`@types/airtable`]. The main difference is `Record` and `RecordData`. [`@types/airtable`] doesn't type Record's methods so it doesn't need to distinguish between data for updating records and `Record` instances that have methods to update that single record.

### `@typescript-eslint/no-explicit-any`

Many source files have any typed arguments and returns. Most of the any typed arguments have been typed with generics. The remaining arguments have been aliased based on what kind of use that argument has. I hope these aliases make further work to type those easier by being able to iterate a search to those instances.

Here is an example from `record.ts` where it has two aliases. The first `RecordError` is used for `error` arguments in callbacks. The second `RecordJson` is for arguments with json results from api calls.

```ts
/* eslint-disable @typescript-eslint/no-explicit-any */
type RecordError = any;
type RecordJson = any;
/* eslint-enable @typescript-eslint/no-explicit-any */
```

### /src/*

The main change for most of these files is the addition of the generic `TFields` that extends `FieldSet`. These names match the same types in [`@types/airtable`].

In addition to the types like in [`@types/airtable`]. Many arguments and variables have also been type and are sometimes casted like in `table.ts`.

### /test/*

I didn't want to change any code in this directory. The only changes are to `/test/.eslintrc` and `/test/test_helpers.js`. Virtually all the warnings were from `no-var` and the remaining warning in `test_helpers` was `no-unused-vars`.

[`@types/airtable`]: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/bacc18855fdebeec1d6edb68677100684254fa46/types/airtable/index.d.ts

### CODE REVIEW

File: src/table.ts [hunk 1]
```diff
@@ -9,57 +9,106 @@ import {QueryParams} from './query_params';
...
-    (recordId: string): Promise<Record>;
-    (recordId: string, done: RecordCallback): void;
+interface TableSelectRecord<TFields extends FieldSet> {
+    (params?: QueryParams<TFields>): Query<TFields>;
 }
 
-interface TableCreateRecords {
-    (recordsData: any[], optionalParameters?: OptionalParameters): Promise<Record[]>;
+interface TableCreateRecords<TFields extends FieldSet> {
+    (recordsData: CreateRecords<TFields>, optionalParameters?: OptionalParameters): Promise<
```
Comment (mzgoddard) [inline src/table.ts:51]: `@types/airtable` types `recordsData` as `TFields[]`. #210 types `recordsData` as `string[]` as an example. [//airtable.com/api] types `recordsData` as `{fields: TFields}[]`.

Are all there right? That is currently what `CreateRecords` expands to: `string[] | TFields[] | {fields: TFields}[]`.

Comment (mzgoddard) [reply src/table.ts:51]: As well should we type with the given `TFields`? Or should we use `Partial<TFields>`? Using `TFields` directly is more flexible for users, they can opt to make the fields optional. Using `Partial<TFields>` users get optional fields but can't force them to be required.

File: src/table.ts [hunk 2]
```diff
@@ -9,57 +9,106 @@ import {QueryParams} from './query_params';
...
+        Record<TFields>
+    >;
+    (
+        recordId: string,
+        recordData: Partial<TFields>,
+        opts: OptionalParameters,
+        done: RecordCallback<TFields>
+    ): void;
+    (recordId: string, recordData: Partial<TFields>, done: RecordCallback<TFields>): void;
+    (recordsData: RecordData<Partial<TFields>>[], opts?: OptionalParameters): Promise<
```
Comment (mzgoddard) [inline src/table.ts:82]: Based on [//airtable.com/api] I believe this is accurate.

File: test/test_helpers.js [hunk 3]
```diff
@@ -112,8 +112,9 @@ function getMockEnvironmentAsync(options) {
     });
 
     // istanbul ignore next
-    // eslint-disable-next-line no-unused-vars
```
Comment (mzgoddard) [inline test/test_helpers.js:115]: `// eslint-disable-next-line no-unused-vars` was not working on my system. `// eslint-disable-next-line` without any rules worked. So I switch to `/* eslint-disable no-unused-vars */` here to be able to disable the specific rule.

**Review by benrazon — APPROVED** [review]

**Comment by replygirl** [comment]
will this be released soon?

**Comment by nickhaughton** [comment]
@benrazon  Will this be released? If not, can we remove the types from the published package? It'd be far easier to use this package if they didn't exist.

**Comment by benrazon** [comment]
Sorry about the delay here. Mzgoddard had one more change that I had reviewed and just wasn't merging cleanly, so I was waiting for him to rebase it before releasing. This ended up taking much longer than I anticipated. I'll hopefully have a release out within a week.

**Pull request merged**